samza-azure/src/main/java/org/apache/samza/system/eventhub/consumer/EventHubSystemConsumer.java

```diff
@@ -0,0 +1,321 @@
+ * │   ┌───────────────────────────────────────┐   │        ┌─────────────────────┐
+ * │   │                                       │   │        │                     │
+ * │   │       PartitionReceiveHandler_1       │───┼───────▶│ SSP1-BlockingQueue  ├──────┐
+ * │   │                                       │   │        │                     │      │
+ * │   └───────────────────────────────────────┘   │        └─────────────────────┘      │
+ * │                                               │                                     │
+ * │   ┌───────────────────────────────────────┐   │        ┌─────────────────────┐      │
+ * │   │                                       │   │        │                     │      │
+ * │   │       PartitionReceiveHandler_2       │───┼───────▶│ SSP2-BlockingQueue  ├──────┤        ┌──────────────────────────┐
+ * │   │                                       │   │        │                     │      ├───────▶│                          │
+ * │   └───────────────────────────────────────┘   │        └─────────────────────┘      └───────▶│  SystemConsumer.poll()   │
+ * │                                               │                                     ┌───────▶│                          │
+ * │                                               │                                     │        └──────────────────────────┘
+ * │                      ...                      │                  ...                │
+ * │                                               │                                     │
+ * │                                               │                                     │
+ * │   ┌───────────────────────────────────────┐   │        ┌─────────────────────┐      │
+ * │   │                                       │   │        │                     │      │
+ * │   │       PartitionReceiveHandler_N       │───┼───────▶│ SSPN-BlockingQueue  ├──────┘
+ * │   │                                       │   │        │                     │
+ * │   └───────────────────────────────────────┘   │        └─────────────────────┘
+ * │                                               │
+ * │                                               │
+ * └───────────────────────────────────────────────┘
+ */
+public class EventHubSystemConsumer extends BlockingEnvelopeMap {
+  private static final Logger LOG = LoggerFactory.getLogger(EventHubSystemConsumer.class);
+  private static final int MAX_EVENT_COUNT_PER_PARTITION_POLL = 50; //TODO
+
+  public static final String START_OF_STREAM = PartitionReceiver.START_OF_STREAM; // -1
+  public static final String END_OF_STREAM = "-2";
+  public static final String AGGREGATE = "aggregate";
+  public static final String EVENT_READ_RATE = "eventReadRate";
+  public static final String EVENT_BYTE_READ_RATE = "eventByteReadRate";
+  public static final String READ_LATENCY = "readLatency";
+  public static final String READ_ERRORS = "readErrors";
+
+  private static Counter aggEventReadRate = null;
```

Comment:
Can these counters be per-instance variables? or do they have to be shared? another nice thing is that you can get rid off this `aggregate` lock.

cc: @lhaiesp to confirm